hnn_core/cell.py

```diff
@@ -696,6 +787,163 @@ def _update_end_pts(self):
+            parent_name = connection[0]
+            parent_loc = connection[1]
+            child_name = connection[2]
+            child_loc = connection[3]
+            parent = (parent_name, parent_loc)
+            child = (child_name, child_loc)
+            child_other_end = (child_name, int(not child_loc))
+
+            # Add the tree entries
+            # If the parent has no child yet
+            if parent not in cell_tree.keys():
+                cell_tree[parent] = list()
+
+            # Connecting child to parent
+            # Note - The coordinates of parent and child node will
+            # be same. Example -
+            # Suppose 0 end of section x is connected to 1 end of soma
+            # then the coordinates of 0 end of section will be same as
+            # coordinates of 1 end of soma.
+            cell_tree[parent].append(child)
+            end_pts_used.add(parent)
+            end_pts_used.add(child_loc)
+
+            # Defining the edge (connect the ends of a section)
+            # Example(Cont.) - The 1 end of section x will be connected
+            # to 0 end of section x. The distance between the two ends
+            # is the length of the section.
+            if child_other_end not in end_pts_used:
+                cell_tree[child] = list()
+                cell_tree[child].append(child_other_end)
+                end_pts_used.add(child_other_end)
+
+        self.cell_tree = cell_tree
+
+    def _update_section_end_pts_L(self, sec_name, end_num, dpt):
+        x = self.sections[sec_name].end_pts[end_num][0]
+        y = self.sections[sec_name].end_pts[end_num][1]
+        z = self.sections[sec_name].end_pts[end_num][2]
+        self.sections[sec_name].end_pts[end_num][0] = x + dpt[0]
+        self.sections[sec_name].end_pts[end_num][1] = y + dpt[1]
+        self.sections[sec_name].end_pts[end_num][2] = z + dpt[2]
+
+        # If current node is a leaf node
+        if (sec_name, end_num) not in self.cell_tree.keys():
+            return
+
+        # If current node is an internal node
+        for (child_name, conn) in self.cell_tree[(sec_name, end_num)]:
+            self._update_section_end_pts_L(child_name, conn, dpt)
+
+    def define_shape(self, sec_name, flag_start):
+        """Redefines end_pts according to section lengths.
+        Detects change in section lengths of the sections in the
+        subtree of the input section.
+
+        Parameters
+        ----------
+        sec_name : Section name
+            end_pts will be modified to adjust changes in length
+            of any section in the subtree of sec_name
+        flag_start : 0 | 1
+            End pt of the section closer to the soma
+
+        Note
+        ----
+        Using sec_name as 'soma' and flag_start as 0 checks for changes
+        in any section length of the cell as (soma, 0) is the root node
+        of the cell.
+        """
+        # Find the end pts of the section
+        flag_end = int(not flag_start)
```

Comment:
Think of making the indexing more explicit, even with if/else statements, rather than boolean logic

Reply:
Added this